storage/src/file_storage.rs

```diff
@@ -0,0 +1,200 @@
+/*
+    Quickwit
+    Copyright (C) 2021 Quickwit Inc.
+
+    Quickwit is offered under the AGPL v3.0 and as commercial software.
+    For commercial licensing, contact us at hello@quickwit.io.
+
+    AGPL:
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use crate::{PutPayload, Storage, StorageErrorKind, StorageFactory, StorageResult};
+use async_trait::async_trait;
+use std::fmt;
+use std::io::SeekFrom;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tokio::fs;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+
+/// File system compatible storage implementation.
+#[derive(Clone)]
+pub struct FileStorage {
+    root: PathBuf,
+}
+
+impl fmt::Debug for FileStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "FileStorage(root={:?})", &self.root)
+    }
+}
+
+impl FileStorage {
+    /// Creates a file storage instance given a root path.
+    pub fn new(root: PathBuf) -> Self {
+        Self { root }
+    }
+
+    /// Creates a file storage instance given a uri
+    pub fn from_uri(uri: &str) -> StorageResult<FileStorage> {
```

Comment:
Im ok with that...
I think I would have prefered something that takes a filepath however, and extract the filepath from the uri in the Factory instead.

Reply:
Ah you also have the FileStorage::new, so no need to change anything I think.